Add audio feedback for changing volume

nekodex opened the pull request:
Also updates `OsuSliderBar` to use the same sample and behave similarly (sound-wise)

- [x] depends on ppy/osu-resources#135


https://user-images.githubusercontent.com/191335/124297489-16430380-db96-11eb-9a1c-42baf2780f27.mp4

nekodex committed:
Add audio feedback to changing volume

nekodex committed:
Update OsuSliderBar to use new notch-tick sample and tweak pitch ramping

peppy committed:
Update resources

bdach commented on osu.Game/Graphics/UserInterface/OsuSliderBar.cs:
```diff
@@ -160,11 +161,11 @@ private void playSample(T value)
             var channel = sample.Play();
 
-            channel.Frequency.Value = 1 + NormalizedValue * 0.2f;
+            channel.Frequency.Value = 1 + NormalizedValue * 0.2f + RNG.NextDouble(0.02f);
```
This is going to be a bit of a nitpick but I'd probably make this

```suggestion
            channel.Frequency.Value = 1 + NormalizedValue * 0.2f + RNG.NextDouble(0.02f) - 0.01f;
```

The reason for that is that `RNG.NextDouble(x)` returns random values from the interval `[0, x]`. So as-is, the distribution of possible frequencies is the continuous range of `[initial_value, initial_value + 0.02f]`. Shifting that back down by `0.01f` yields a distribution of `[initial_value - 0.01f, initial_value + 0.01f]` and so the distribution is centered around the initial value.

peppy replied:
The change you're proposing will actually change the audible sound. @nekodex can you confirm you only intended the random portion to be in an upwards direction here?

nekodex replied:
The goal here was just to add a tiny bit of randomisation to the otherwise linear pitch scaling to make things sound a bit more natural, the range of that randomisation was less important.

I'll try the suggested change, but honestly I'm not sure it would result in an appreciable difference to the listener?

bdach replied:
It won't be audible for sure, the current code was just sticking out to my inner sense of symmetry, which is why I pre-empted the comment by saying it's a nitpick.

`HoverSounds` already does things in a way similar to what I proposed, for whatever that's worth:

https://github.com/ppy/osu/blob/c2759121b9fb6acafd37d5a72720a1bef0fc1ccf/osu.Game/Graphics/UserInterface/HoverSounds.cs#L39

peppy replied:
@bdach did you mean to adjust the initial value to keep the same range?

bdach replied:
No, my intention was that range would change. Compare the `HoverSounds` snippet above, which uses a range of [0.98, 1.02] for frequency, which is centered at 1. What this pull does right now would be more akin to [1, 1.04]. But honestly not really even fussed about this that much anymore.

bdach commented on osu.Game/Graphics/UserInterface/OsuSliderBar.cs:
```diff
@@ -160,11 +161,11 @@ private void playSample(T value)
+            channel.Frequency.Value = 1 + NormalizedValue * 0.2f + RNG.NextDouble(0.02f);
             if (NormalizedValue == 0)
-                channel.Frequency.Value -= 0.4f;
+                channel.Frequency.Value -= 0.5f;
             else if (NormalizedValue == 1)
-                channel.Frequency.Value += 0.4f;
+                channel.Frequency.Value -= 0.5f;
```
Was this really meant to be a `-=` in both cases? From this diff alone it half-looks like an error/oversight, because the signs were different before, and because the conditions aren't merged which would make sense if the two cases were to be treated in the same way.

peppy replied:
This was indeed intentional. Maybe worth a comment attached to it since it doesn't match expectations when reading the code.

nekodex replied:
Was intentional, yeah. I kinda overlooked merging the conditionals as I was originally just messing around with numbers until it sounded good.

bdach commented on osu.Game/Overlays/Volume/VolumeMeter.cs:
```diff
@@ -217,6 +227,29 @@ protected double DisplayVolume
+                {
+                    const int sfx_debounce_time = 30;
+
+                    if (firstUpdate ||
+                        !displayVolumeChanged ||
+                        Time.Current - sampleLastPlaybackTime <= sfx_debounce_time)
+                    {
+                        firstUpdate = false;
+                        return;
+                    }
```
Dunno if it's just me but I find combining the `firstUpdate` logic with general "should play this sample" logic a little overcomplicated. I'd probably prefer writing two separate guards:

```csharp
if (firstUpdate)
{
    firstUpdate = false;
    return;
}

if (!displayVolumeChanged || Time.Current - sampleLastPlaybackTime <= sfx_debounce_time)
{
    return;
}
```

I am guessing a bit as to what the intent here was, so this may not be right if I misread.

nekodex replied:
Yeah, that was the intent.

I originally wrote it as two separate guards, but combined them to avoid the double returns. If you think it's more readable with them separated, then I have no issue doing that.

peppy replied:
I'm rewriting this function as it's all a bit weird.

peppy replied:
I've rewritten this function. Probably worth someone else double-checking the new logic. Hopefully reads better.

peppy replied:
Or I haven't because changes were pushed ahead of mine..

nekodex replied:
Ah I didn't notice the message before pushing, wasn't expecting you to jump in and rewrite 🤷

peppy committed:
Merge branch 'master' into notch-tick-sfx

peppy committed:
Tidy up weird bind logic

nekodex committed:
Center pitch randomisation around base pitch

peppy committed:
Remove usage of `bool` for initial playback

Also refactors the whole method to generally clean things up. One more
important fix is setting the frequency on the channel before starting
playback, so avoid the frequency potentially being adjusted after the
playback is already started.

peppy committed:
Add note about intentional downward pitch, against expectations

peppy committed:
Fix one more instance of incorrect playback/frequency set order

peppy commented:
One thing to note is that in cases you're changing things like the playback frequency, make sure to use `sample.GetChannel()` and change the frequency before calling `Play()`. Due to the multithreaded nature it can happen that the playback occurs before the adjustments are applied otherwise.

peppy committed:
Merge conditionals in line with other case of same logic

nekodex commented:
> One thing to note is that in cases you're changing things like the playback frequency, make sure to use `sample.GetChannel()` and change the frequency before calling `Play()`. Due to the multithreaded nature it can happen that the playback occurs before the adjustments are applied otherwise.

I kinda suspected as much, but figured it was fine as that's how it was already being done in `OsuSliderBar.cs` before my changes.

smoogipoo committed:
Merge branch 'master' into notch-tick-sfx
